witchcraft-log/src/mdc.rs

```diff
@@ -249,3 +269,59 @@ impl Snapshot {
         &mut self.unsafe_
     }
 }
+
+/// A guard object which resets the MDC to an earlier state when it drops.
+pub struct Scope {
+    old: Snapshot,
+}
+
+impl Drop for Scope {
+    fn drop(&mut self) {
+        swap(&mut self.old);
+    }
+}
+
+/// A future which manages the MDC across polls to a delegate.
+#[pin_project(PinnedDrop)]
+pub struct Bind<F> {
+    #[pin]
+    future: Option<F>,
+    snapshot: Snapshot,
+}
+
+#[pinned_drop]
+impl<F> PinnedDrop for Bind<F> {
+    fn drop(self: Pin<&mut Self>) {
+        let mut this = self.project();
+        let _guard = scope_with(this.snapshot);
+        this.future.set(None);
+    }
+}
+
+impl<F> Future for Bind<F>
+where
+    F: Future,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let _guard = scope_with(this.snapshot);
+        this.future.as_pin_mut().unwrap().poll(cx)
+    }
+}
+
+fn scope_with(snapshot: &mut Snapshot) -> ScopeWith<'_> {
+    swap(snapshot);
+    ScopeWith { snapshot }
+}
+
+struct ScopeWith<'a> {
```

Comment:
Can you clarify with a doc the purpose of this vs. Scope?

Reply:
Updated. The difference is that Scope works starts with the current MDC, while scope_with starts with the provided snapshot state.